api/src/test/kotlin/org/jetbrains/kotlinx/dl/api/core/layer/LayerTest.kt

```diff
@@ -0,0 +1,142 @@
+
+import org.jetbrains.kotlinx.dl.api.core.KGraph
+import org.jetbrains.kotlinx.dl.api.core.shape.*
+import org.junit.jupiter.api.Assertions
+import org.tensorflow.*
+import org.tensorflow.op.Ops
+import kotlin.IllegalArgumentException
+
+enum class RunMode {
+    EAGER,
+    GRAPH
+}
+
+open class LayerTest {
+
+    private fun getInputOp(tf: Ops, input: Array<*>): Operand<Float> {
+        return when (input.shape.numDimensions()) {
+            1 -> tf.constant(input.map { it as Float }.toFloatArray())
+            2 -> tf.constant(input.cast2D<FloatArray>())
+            3 -> tf.constant(input.cast3D<FloatArray>())
+            4 -> tf.constant(input.cast4D<FloatArray>())
+            5 -> tf.constant(input.cast5D<FloatArray>())
+            else -> throw IllegalArgumentException("Inputs with more than 5 dimensions are not supported yet!")
+        }
+    }
+
+    private fun getLayerOutputOp(
+        tf: Ops,
+        layer: Layer,
+        input: Array<*>,
+        kGraph: KGraph,
+    ) : Output<*> {
+        val inputShape = input.shape
+        layer.build(tf, kGraph, inputShape)
+        val inputOp = getInputOp(tf, input)
+        val isTraining = tf.constant(true)
+        val numberOfLosses = tf.constant(1.0f)
+        val output = layer.forward(tf, inputOp, isTraining, numberOfLosses).asOutput()
+        return output
+    }
+
+    protected fun runLayerInEagerMode(
+        layer: Layer,
+        input: Array<*>,
+    ): Tensor<*> {
+        EagerSession.create().use {
+            val tf = Ops.create()
+            val kGraph = KGraph(Graph().toGraphDef())
+            val outputOp = getLayerOutputOp(tf, layer, input, kGraph)
+            return outputOp.tensor()
+        }
+    }
+
+    protected fun runLayerInGraphMode(
+        layer: Layer,
+        input: Array<*>,
+    ) : Tensor<*> {
+        Graph().use { graph ->
+            Session(graph).use { session ->
+                val tf = Ops.create(graph)
+                KGraph(graph.toGraphDef()).use { kGraph ->
+                    val outputOp = getLayerOutputOp(tf, layer, input, kGraph)
+                    /**
+                     * This check is necessary, otherwise it would throw the following error
+                     * if the layer has no trainable parameters:
+                     * ```
+                     * Must specify at least one target to fetch or execute.
+                     * ```
+                     * However, using `layer.weights.isNotEmpty()` for the condition would not work
+                     * because according to the current implementation of `extractWeights` the layer
+                     * should belong to an actual model and therefore it throws an error. Hence, we
+                     * are using `paramCount` instead to determine whether the layer has any trainable
+                     * weights.
+                     */
+                    if (layer.paramCount > 0)
+                        kGraph.initializeGraphVariables(session)
```

Comment:
Why does `initializeGraphVariables` always expect some variables to be created and present? Couldn't it be modified so that it does not run the session if there is no variable to initialize? In that case, we can get rid of this `if` statement.

Reply:
It's a good idea, please create an enhancement about it, I will handle your idea and answer there